File: container/skills/programmatic-mcp/scripts/call_tool.js
#!/usr/bin/env node
/**
 * Call a single MCP tool programmatically using Node.js.
 *
 * Usage:
 *   node call_tool.js <tool_name> '<json_args>' -- <command> [args...]
 *
 * Examples:
 *   node call_tool.js read_file '{"path": "/tmp/test.txt"}' -- npx -y @modelcontextprotocol/server-filesystem /tmp
 *   node call_tool.js list_directory '{"path": "/tmp"}' -- npx -y @modelcontextprotocol/server-filesystem /tmp
 */

import { Client } from "@modelcontextprotocol/sdk/client/index.js";
import { StdioClientTransport } from "@modelcontextprotocol/sdk/client/stdio.js";

async function callTool(toolName, args, command, commandArgs) {
  const transport = new StdioClientTransport({
    command,
    args: commandArgs,
  });

  const client = new Client({ name: "call-tool-script", version: "1.0.0" });
  await client.connect(transport);

  try {
    const result = await client.callTool({ name: toolName, arguments: args });
    for (const content of result.content) {
      if (content.type === "text") {
        console.log(content.text);
      } else {
        console.log(JSON.stringify(content, null, 2));
      }
    }
  } finally {
    await client.close();
  }
}

// Parse CLI args
const argv = process.argv.slice(2);
const sepIndex = argv.indexOf("--");

if (sepIndex === -1 || sepIndex < 2) {
  console.log("Usage: node call_tool.js <tool_name> '<json_args>' -- <command> [args...]");
  process.exit(1);
}

const toolName = argv[0];
const toolArgs = JSON.parse(argv[1]);
const command = argv[sepIndex + 1];
const commandArgs = argv.slice(sepIndex + 2);

callTool(toolName, toolArgs, command, commandArgs).catch((err) => {
  console.error("Error:", err.message);
  process.exit(1);
});

File: container/skills/programmatic-mcp/scripts/call_tool.py
#!/usr/bin/env python3
"""Call a single MCP tool programmatically.

Usage:
    python call_tool.py <tool_name> '<json_args>' -- <command> [args...]

Examples:
    python call_tool.py read_file '{"path": "/tmp/test.txt"}' -- npx -y @modelcontextprotocol/server-filesystem /tmp
    python call_tool.py list_directory '{"path": "/tmp"}' -- npx -y @modelcontextprotocol/server-filesystem /tmp
"""

import asyncio
import json
import sys

from mcp import ClientSession, StdioServerParameters
from mcp.client.stdio import stdio_client


async def call_tool(tool_name: str, arguments: dict, command: str, args: list[str]):
    server_params = StdioServerParameters(command=command, args=args, env=None)

    async with stdio_client(server_params) as (read, write):
        async with ClientSession(read, write) as session:
            await session.initialize()

            result = await session.call_tool(tool_name, arguments=arguments)

            if result.isError:
                print(f"ERROR: Tool call failed", file=sys.stderr)
                for content in result.content:
                    text = content.text if hasattr(content, "text") else str(content)
                    print(text, file=sys.stderr)
                sys.exit(1)

            for content in result.content:
                if hasattr(content, "text"):
                    print(content.text)
                else:
                    print(json.dumps(content.model_dump(), indent=2))


def main():
    if "--" not in sys.argv:
        print(__doc__)
        sys.exit(1)

    sep = sys.argv.index("--")
    before = sys.argv[1:sep]
    after = sys.argv[sep + 1 :]

    if len(before) < 2 or len(after) < 1:
        print(__doc__)
        sys.exit(1)

    tool_name = before[0]
    arguments = json.loads(before[1])
    command = after[0]
    args = after[1:]

    asyncio.run(call_tool(tool_name, arguments, command, args))


if __name__ == "__main__":
    main()
